eval/eval.go

```diff
@@ -932,6 +969,87 @@ func (e *evalContext) evaluateBuiltinOpen(x *expr, repr *openExpr) *value {
 	return unexport(output, x)
 }
 
+// evaluateBuiltinOpen evaluates a call to the fn::rotate builtin.
+func (e *evalContext) evaluateBuiltinRotate(x *expr, repr *rotateExpr) *value {
+	v := &value{def: x}
+
+	// Can happen if there are parse errors.
+	if repr.node.Rotator == nil {
+		v.schema = schema.Always()
+		v.unknown = true
+		return v
+	}
+
+	rotator, err := e.providers.LoadRotator(e.ctx, repr.node.Rotator.GetValue())
+	if err != nil {
+		e.errorf(repr.syntax(), "%v", err)
+	} else {
+		inputSchema, stateSchema, outputSchema := rotator.Schema()
+		if err := inputSchema.Compile(); err != nil {
+			e.errorf(repr.syntax(), "internal error: invalid input schema (%v)", err)
+		} else {
+			repr.inputSchema = inputSchema
+		}
+		if err := stateSchema.Compile(); err != nil {
+			e.errorf(repr.syntax(), "internal error: invalid state schema (%v)", err)
+		} else {
+			repr.stateSchema = stateSchema
+		}
+		if err := outputSchema.Compile(); err != nil {
+			e.errorf(repr.syntax(), "internal error: invalid schema (%v)", err)
+		} else {
+			x.schema = outputSchema
+		}
+	}
+	v.schema = x.schema
+
+	inputs, inputsOK := e.evaluateTypedExpr(repr.inputs, repr.inputSchema)
+	state, stateOK := e.evaluateTypedExpr(repr.state, repr.stateSchema)
+	if !inputsOK || inputs.containsUnknowns() || !stateOK || state.containsUnknowns() || e.validating || err != nil {
+		v.unknown = true
+		return v
+	}
+
+	// if rotating, invoke prior to open
+	if e.rotating {
+		newState, err := rotator.Rotate(
+			e.ctx,
+			inputs.export("").Value.(map[string]esc.Value),
+			state.export("").Value.(map[string]esc.Value),
+			e.execContext,
+		)
+		if err != nil {
+			e.errorf(repr.syntax(), "rotate: %s", err.Error())
+			v.unknown = true
+			return v
+		}
+
+		// todo: validate newState conforms to state schema
+
+		e.patchOutputs = append(e.patchOutputs, &Patch{
+			// rotation output is written back to the fn's `state` input
+			DocPath:     x.path + "." + repr.node.Name().GetValue() + ".state",
```

Comment:
probably safer to do something like this, which will apply any necessary escaping to `repr.node.Name().GetValue()`:

```suggestion
			DocPath:     util.JoinKey(x.path, repr.node.Name().GetValue()) + ".state",
```